### PR TITLE
feat: HIP-1046 

### PR DESCRIPTION
**Pull request by ivaylonikolov7**
### Added
- `WebClient.forMirrorNetwork()` factory method now works and correctly initializes and updates the address book during initialization.
- `WebClient` now supports scheduled network updates via the `scheduleNetworkUpdate: true` option, and manual updates using the `updateNetwork()` method.
- `AddressBookQuery` is now supported in browser environments.
- `MirrorNodeContractCallQuery` is now supported in browser environments.
- `MirrorNodeContractЕstimateQuery` is now supported in browser environments.
**Related issue(s)**:
[3102](https://github.com/hiero-ledger/hiero-sdk-js/issues/3102)
[3148
](https://github.com/hiero-ledger/hiero-sdk-js/issues/3148)

Fixes #
[3102](https://github.com/hiero-ledger/hiero-sdk-js/issues/3102)
[3148
](https://github.com/hiero-ledger/hiero-sdk-js/issues/3148)

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/network/AddressBookQueryWeb.js [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+ *   address: string,
+ *   service_endpoints: {
+ *     domain_name: string,
+ *     ip_address_v4: string,
+ *     port: number
+ *   }[],
+ *   description: string,
+ *   stake: number
+ * }>} nodes
+ */
```
Comment (ivaylogarnev-limechain) [inline src/network/AddressBookQueryWeb.js:46]: Don't we have interface for that?

File: src/channel/WebChannel.js [hunk 2]
```diff
@@ -39,11 +39,14 @@ export default class WebChannel extends Channel {
         // eslint-disable-next-line @typescript-eslint/no-misused-promises
         return async (method, requestData, callback) => {
             try {
+                const address = this._address.includes("http")
+                    ? this._address
+                    : `http://${this._address}`;
```
Comment (venilinvasilev) [inline src/channel/WebChannel.js]: What do you think about this should we instead add a check if this is in local environment to add http and use https if not?

File: src/client/WebClient.js [hunk 3]
```diff
@@ -156,10 +160,43 @@ export default class WebClient extends Client {
...
+        for (const nodeAddress of addressBook.nodeAddresses) {
+            for (const endpoint of nodeAddress.addresses) {
+                if (nodeAddress.accountId != null) {
+                    console.log({ endpoint });
+                    network[endpoint.toString()] = nodeAddress.accountId;
+                }
+            }
+        }
+
+        client.setNetwork(network);
```
Comment (venilinvasilev) [inline src/client/WebClient.js]: I cannot use client.network.setNetworkFromAddressBook because of underlying logic which strips node addresses with ports different than 50211 and 50212.
The approach proposed from @ivaylonikolov7 with branching the Network for browser env as WebNetwork is not applicable and won't work in this case so that's why I've made the network mapping here. WDYT?

File: src/client/WebClient.js [hunk 4]
```diff
@@ -196,6 +233,46 @@ export default class WebClient extends Client {
...
+
+            for (const nodeAddress of addressBook.nodeAddresses) {
+                for (const endpoint of nodeAddress.addresses) {
+                    if (nodeAddress.accountId != null) {
+                        network[endpoint.toString()] = nodeAddress.accountId;
+                    }
+                }
+            }
+
+            this.setNetwork(network);
```
Comment (venilinvasilev) [inline src/client/WebClient.js]: Same concern as my above comment.

File: src/client/WebClient.js [hunk 5]
```diff
@@ -185,17 +205,75 @@ export default class WebClient extends Client {
...
+    async updateNetwork() {
+        if (this._isUpdatingNetwork) {
+            return;
+        }
+
+        this._isUpdatingNetwork = true;
+
+        try {
+            const addressBook = await new AddressBookQuery()
+                .setFileId(FileId.ADDRESS_BOOK)
```
Comment (0xivanov) [inline src/client/WebClient.js]: FYI, This would need to use the shard and realms

File: src/network/AddressBookQueryWeb.js [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+                        } during attempt ${
+                            this._attempt
+                        }. Waiting ${delay} ms before next attempt: ${message}`,
+                    );
+                }
+
+                this._attempt += 1;
+
+                setTimeout(() => {
+                    void this._makeFetchRequest(
```
Comment (0xivanov) [inline src/network/AddressBookQueryWeb.js]: I don't really like the idea to make this recursive. A simple for loop using the attemps as index would do the job.

File: src/query/MirrorNodeContractEstimateQuery.js [hunk 7]
```diff
@@ -12,7 +12,7 @@ import MirrorNodeContractQuery from "./MirrorNodeContractQuery.js";
  * to execute the same call on the main network. It's useful for determining the appropriate
  * gas limit before submitting an actual transaction.
  */
-export default class MirrorNodeContractCallQuery extends MirrorNodeContractQuery {
+export default class MirrorNodeContractEstimateQuery extends MirrorNodeContractQuery {
```
Comment (0xivanov) [inline src/query/MirrorNodeContractEstimateQuery.js:15]: interesting.. and nice catch

File: src/client/WebClient.js (hunk 5 above)
Comment (venilinvasilev) [reply src/client/WebClient.js]: Thanks for flagging this. It uses now.

File: src/address_book/IPv4Address.js [hunk 8]
```diff
@@ -90,6 +90,28 @@ export default class IPv4Address {
...
+     */
+    static _fromString(address) {
+        const parts = address.split(".");
+        if (parts.length !== 4) {
+            throw new Error("Invalid IPv4 address format");
+        }
+
+        const network = new IPv4AddressPart()
+            .setLeft(parseInt(parts[0], 10))
+            .setRight(parseInt(parts[1], 10));
```
Comment (ivaylonikolov7) [inline src/address_book/IPv4Address.js]: isn't 10 the default one?

File: src/client/Client.js [hunk 9]
```diff
@@ -130,7 +130,6 @@ export default class Client {
         this._requestTimeout = null;
 
         /**
-         * @private
```
Comment (ivaylonikolov7) [inline src/client/Client.js:133]: Why expose this? And if we are going to expose this wouldn't it be better to do so by using a getter?

File: src/native.js [hunk 10]
```diff
@@ -7,3 +7,4 @@
 export * from "./exports.js";
 
 export { default as Client } from "./client/NativeClient.js";
+export { default as AddressBookQuery } from "./network/AddressBookQuery.js";
```
Comment (ivaylonikolov7) [inline src/native.js:10]: Did some quick search and found that react native doesn't support gRPC requests too. Can we have `AddressBookQueryWeb` here too?

File: src/channel/WebChannel.js (hunk 2 above)
Comment (ivaylonikolov7) [reply src/channel/WebChannel.js]: I think this is fine. Local environment will always use http, and production will use https.

File: src/constants/ClientConstants.js [hunk 11]
```diff
@@ -3,63 +3,63 @@ import AccountId from "../account/AccountId.js";
...
+    "node06.swirldslabs.com:443": new AccountId(9),
+    "node07.swirldslabs.com:443": new AccountId(10),
+    "node09.swirldslabs.com:443": new AccountId(12),
+    "node10.swirldslabs.com:443": new AccountId(13),
+    "node11.swirldslabs.com:443": new AccountId(14),
+    "node12.swirldslabs.com:443": new AccountId(15),
+    "node14.swirldslabs.com:443": new AccountId(17),
+    "node15.swirldslabs.com:443": new AccountId(18),
+    "node16.swirldslabs.com:443": new AccountId(19),
+    "node17.swirldslabs.com:443": new AccountId(20),
```
Comment (ivaylonikolov7) [inline src/constants/ClientConstants.js:21]: Can we make it to keep the `https://` in front of the address. I am worried that this might break the SDK somewhere else. I remember you had some issues with the `fetch` basically calling `http://http://" but seems like we could create some unnecessary headaches for us with this change.

File: src/native.js (hunk 10 above)
Comment (venilinvasilev) [reply src/native.js:10]: Thanks for researching and flagging this. I think this would be great addition, but I would separate this change from this PR as it is quite big already. Furthermore this needs to be tested on Native environment which would slow down the process of getting these changes to being merged. WDYT? Would you like to create an issue for this?

File: src/constants/ClientConstants.js (hunk 11 above)
Comment (venilinvasilev) [reply src/constants/ClientConstants.js:21]: We need to keep these like this, because this is the way that we are getting the addresses/domains from the address book query (without the "https://" schema). If I leave them here then I need to add some checks to transform the data for when we set the list from the hardcoded values including the schema and not when we are getting the list from the addressbok query response. Lets keep these consistent I will test this thoroughly.

File: src/network/AddressBookQueryWeb.js (hunk 1 above)
Comment (venilinvasilev) [reply src/network/AddressBookQueryWeb.js]: No. This is the response schema defined from the mirror rest API

File: src/address_book/IPv4Address.js (hunk 8 above)
Comment (venilinvasilev) [reply src/address_book/IPv4Address.js]: Actually I will use Number instead

File: src/client/Client.js (hunk 9 above)
Comment (venilinvasilev) [reply src/client/Client.js:133]: We're exposing this internally for the sub-classes not for the end-user that's why it starts with "_"

File: src/network/AddressBookQueryWeb.js (hunk 6 above)
Comment (venilinvasilev) [reply src/network/AddressBookQueryWeb.js]: Removed recursion

File: examples/frontend-examples/src/app/client/grpc-web-proxy/page.js [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+    const [clientNetwork, setClientNetwork] = useState(initialNetwork);
+    const [isLoading, setIsLoading] = useState(false);
+    const [message, setMessage] = useState("");
+    const [messageType, setMessageType] = useState("info");
+    const [createdAccount, setCreatedAccount] = useState(null);
+
+    const client = useMemo(() => {
+        return WebClient.forNetwork(initialNetwork).setOperator(
+            AccountId.fromString(process.env.NEXT_PUBLIC_OPERATOR_ID),
+            PrivateKey.fromStringDer(process.env.NEXT_PUBLIC_OPERATOR_KEY),
```
Comment (ivaylonikolov7) [inline examples/frontend-examples/src/app/client/grpc-web-proxy/page.js:44]: We already have ` PrivateKey.fromStringED25519(process.env.NEXT_PUBLIC_OPERATOR_KEY),`

Are we going to stick to the already implemented way of setting up the client, wdyt? 

I can see your point that by using `fromStringDer` we could use both ECDSA and ED25519 key but we already have implemented one way. 

My opinion is that your way of handling this is better but we might have to change the setting of the client in the other page too in this PR?

**Review by venilinvasilev — DISMISSED** [review]

**Comment by venilinvasilev** [comment]
LGTM

File: src/client/WebClient.js [hunk 13]
```diff
@@ -201,17 +215,73 @@ export default class WebClient extends Client {
...
+                    break;
+                case "testnet":
+                    this._mirrorNetwork.setNetwork(MirrorNetwork.TESTNET);
+                    break;
+                case "mainnet":
+                    this._mirrorNetwork.setNetwork(MirrorNetwork.MAINNET);
+                    break;
+                default:
+                    this._mirrorNetwork.setNetwork([mirrorNetwork]);
+            }
```
Comment (0xivanov) [inline src/client/WebClient.js:235]: Can you extract this logic in the parent `Client.js`?

File: src/network/AddressBookQueryWeb.js [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+ * or from the most recent address book if no file ID is specified. The response
+ * contains node metadata including IP addresses and ports for both node and mirror
+ * node services.
+ * @augments {Query<NodeAddressBook>}
+ */
+export default class AddressBookQueryWeb extends Query {
+    /**
+     * @param {object} props
+     * @param {FileId | string} [props.fileId]
+     * @param {number} [props.limit]
```
Comment (0xivanov) [inline src/network/AddressBookQueryWeb.js:68]: Could document what this limit does. Here and in the setter/getter.

File: src/network/AddressBookQueryWeb.js [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+                        );
+                    }
+
+                    // Wait before next attempt
+                    await new Promise((resolve) => setTimeout(resolve, delay));
+                    continue;
+                }
+
+                // If we shouldn't retry or have exhausted attempts, reject
+                reject(new Error("failed to query address book"));
```
Comment (0xivanov) [inline src/network/AddressBookQueryWeb.js]: Could specify that we reached max attempts

File: test/unit/browser/WebClient.js [hunk 16]
```diff
@@ -0,0 +1,1560 @@
...
+            const updatedNetwork = client.network;
+
+            const initialEntries = createNetworkAddressNodeSet(initialNetwork);
+            const updatedEntries = createNetworkAddressNodeSet(updatedNetwork);
+
+            expect(initialEntries).not.toEqual(updatedEntries);
+            expect(updatedEntries.size).toBe(0);
+        });
+
+        it("should change network to empty when mirror response has only nodes with empty domain_name in grpc_proxy_endpoint", async function () {
```
Comment (0xivanov) [inline test/unit/browser/WebClient.js:1467]: Add a test where we have `forNetwork` and have response containing 1 node with valid proxy and 1 node with null/empty domain name proxy.
What would happen?

Comment (0xivanov) [reply test/unit/browser/WebClient.js:1467]: I think the result should be 1 node in the network. Could be tested with `forMainnet/Testnet`,etc.

File: src/client/WebClient.js (hunk 13 above)
Comment (venilinvasilev) [reply src/client/WebClient.js:235]: Good observation. Seems that NativeClient still has nothing to do with MirrorNetwork.
I would extract this when we add MirrorNetwork support to the NativeClient (maybe we need an issue for this).

File: test/unit/browser/WebClient.js (hunk 16 above)
Comment (venilinvasilev) [reply test/unit/browser/WebClient.js:1467]: Nice catch. Seems that indeed I have missed this one!

Comment (venilinvasilev) [reply test/unit/browser/WebClient.js:1467]: It should have them now.

File: src/client/WebClient.js (hunk 13 above)
Comment (venilinvasilev) [reply src/client/WebClient.js:235]: I've created an issue for this: https://github.com/hiero-ledger/hiero-sdk-js/issues/3205

**Review by 0xivanov — APPROVED** [review]
LGTM

**Review by 0xivanov — COMMENTED** [review]
Do not merge

**Review by 0xivanov — APPROVED** [review]
